pageserver/src/layered_repository.rs

```diff
@@ -126,86 +128,61 @@ pub struct Repository {
 impl Repository {
     /// Get Timeline handle for given zenith timeline ID.
     /// This function is idempotent. It doesn't change internal state in any way.
-    pub fn get_timeline(&self, timelineid: ZTimelineId) -> Option<RepositoryTimeline<Timeline>> {
-        self.timelines
-            .lock()
-            .unwrap()
-            .get(&timelineid)
-            .cloned()
-            .map(RepositoryTimeline::from)
-    }
-
-    /// Get Timeline handle for locally available timeline. Load it into memory if it is not loaded.
-    pub fn get_timeline_load(&self, timeline_id: ZTimelineId) -> Result<Arc<Timeline>> {
-        let mut timelines = self.timelines.lock().unwrap();
-        match self.get_timeline_load_internal(timeline_id, &mut timelines)? {
-            Some(local_loaded_timeline) => Ok(local_loaded_timeline),
-            None => anyhow::bail!("cannot get local timeline, unknown timeline id: {timeline_id}"),
-        }
+    pub fn get_timeline(&self, timeline_id: ZTimelineId) -> Option<Arc<Timeline>> {
+        self.timelines.lock().unwrap().get(&timeline_id).cloned()
     }
 
     /// Lists timelines the repository contains.
     /// Up to repository's implementation to omit certain timelines that ar not considered ready for use.
-    pub fn list_timelines(&self) -> Vec<(ZTimelineId, RepositoryTimeline<Timeline>)> {
+    pub fn list_timelines(&self) -> Vec<(ZTimelineId, Arc<Timeline>)> {
         self.timelines
             .lock()
             .unwrap()
             .iter()
-            .map(|(timeline_id, timeline_entry)| {
-                (
-                    *timeline_id,
-                    RepositoryTimeline::from(timeline_entry.clone()),
-                )
-            })
+            .map(|(timeline_id, timeline_entry)| (*timeline_id, Arc::clone(timeline_entry)))
             .collect()
     }
 
     /// Create a new, empty timeline. The caller is responsible for loading data into it
     /// Initdb lsn is provided for timeline impl to be able to perform checks for some operations against it.
     pub fn create_empty_timeline(
         &self,
-        timeline_id: ZTimelineId,
+        new_timeline_id: ZTimelineId,
         initdb_lsn: Lsn,
     ) -> Result<Arc<Timeline>> {
-        let mut timelines = self.timelines.lock().unwrap();
-        let vacant_timeline_entry = match timelines.entry(timeline_id) {
-            Entry::Occupied(_) => bail!("Timeline already exists"),
-            Entry::Vacant(vacant_entry) => vacant_entry,
-        };
+        anyhow::ensure!(
```

Comment:
I've realised, that I've done the same error in `branch_timeline`, thank you for spotting.

The tombstone idea is interesting, yet I think I'd leave the locks for now, since the situation with the PR becomes messy 🙂 

I've addressed all the feedback in https://github.com/neondatabase/neon/pull/2324, where this PR got merged to.